hudi-client/src/main/java/org/apache/hudi/index/HoodieSimpleIndex.java

```diff
@@ -0,0 +1,202 @@
+
+    jsc.setJobGroup(this.getClass().getSimpleName(), "Tagging incoming records with record location");
+    JavaRDD<Tuple2<HoodieKey, Tuple2<HoodieRecord, Option<HoodieRecordLocation>>>> untaggedRecordsRDD = incomingRecords.leftOuterJoin(existingRecords)
+        .map(entry -> new Tuple2(entry._1, new Tuple2(entry._2._1, Option.ofNullable(entry._2._2.orNull()))));
+
+    JavaRDD<HoodieRecord<T>> taggedRecordRDD = untaggedRecordsRDD.map(entry -> getTaggedRecord(entry._2._1, entry._2._2));
+
+    if (config.getSimpleIndexUseCaching()) {
+      recordRDD.unpersist(); // unpersist the input Record RDD
+    }
+    return taggedRecordRDD;
+  }
+
+  /**
+   * Returns an RDD mapping each HoodieKey with a partitionPath/fileID which contains it. Option.Empty if the key is not.
+   * found.
+   *
+   * @param hoodieKeys  keys to lookup
+   * @param jsc         spark context
+   * @param hoodieTable hoodie table object
+   */
+  @Override
+  public JavaPairRDD<HoodieKey, Option<Pair<String, String>>> fetchRecordLocation(JavaRDD<HoodieKey> hoodieKeys,
+                                                                                  JavaSparkContext jsc, HoodieTable<T> hoodieTable) {
+    JavaPairRDD<HoodieKey, Option<HoodieRecordLocation>> incomingRecords =
+        hoodieKeys.mapToPair(entry -> new Tuple2<>(entry, Option.empty()));
+
+    JavaPairRDD<HoodieKey, HoodieRecordLocation> existingRecords = fetchRecordLocations(hoodieKeys, jsc, hoodieTable);
+
+    jsc.setJobGroup(this.getClass().getSimpleName(), "Joining existing records with incoming keys to find record locations");
+    return incomingRecords.leftOuterJoin(existingRecords)
+        .mapToPair(entry -> new Tuple2(entry._1, Option.ofNullable(entry._2._2.orNull())));
+  }
+
+  /**
+   * Fetch record locations for passed in {@link HoodieKey}s.
+   *
+   * @param hoodieKeys  {@link JavaRDD} of {@link HoodieKey}s for which locations are fetched
+   * @param jsc         instance of {@link JavaSparkContext} to use
+   * @param hoodieTable instance of {@link HoodieTable} of interest
+   * @return {@link JavaPairRDD} of {@link HoodieKey} and {@link HoodieRecordLocation}
+   */
+  private JavaPairRDD<HoodieKey, HoodieRecordLocation> fetchRecordLocations(JavaRDD<HoodieKey> hoodieKeys, JavaSparkContext jsc, HoodieTable hoodieTable) {
+
+    JavaPairRDD<String, String> partitionRecordKeyPairRDD = hoodieKeys.mapToPair(entry -> new Tuple2(entry.getPartitionPath(), entry.getRecordKey()));
+    jsc.setJobGroup(this.getClass().getSimpleName(), "Fetching fileInfos for partitions of interest");
+    List<String> affectedPartitionPathList = partitionRecordKeyPairRDD.map(tuple -> tuple._1).distinct().collect();
+    JavaRDD<Tuple2<String, String>> fileInfoList = jsc.parallelize(
+        loadAllFilesForPartitions(affectedPartitionPathList, jsc, hoodieTable)).sortBy(Tuple2::_1, true, config.getSimpleIndexParallelism());
```

Comment:
If you have more than 100 executors , then your performance will be limited by parallelism.. I think we should have simpleIndexParallelism default to 0. if its 0, we have parallelism=number of files, if its set to a non-zero value, we use that ?

Reply:
got it.